### ISSUE TITLE
Performance: watch, key-space-size, watcher-size, put-rate

### ISSUE THREAD

**Issue by gyuho** [coreos/etcd#4686]
Following https://github.com/coreos/etcd/pull/4675.

Let's assume two cases:
1. 1 key, 1000 watchers
2. 1000 different keys, 1000 watchers

Case 1 needs only 1 `Put` request to trigger all 1000 watchers.
Case 2 needs 1000 `Put` requests to fire each of 1000 watchers.

So theoretically, case 1 should be 1000x in terms of through-put.
But in real-world, we have to consider other factors. With very high put-rate,
etcd watchers take more time to process events: case 1 still performs
better, but not as much as 1000x compared to case 2.

- relation between `key-space-size`, and watch performance
- relations between `key-space-size`, `put-rate`, and watch performance

**Comment by gyuho** [comment]
Given other configurations are the same, `1 / key-space-size` is proportional to watch performance:

```
# key-space-size=1000
benchmark --conns=5 --clients=150 \
	watch --watchers=10000 --total=50000 --watched-key-total=10000 \
	--put-rate=1000 --put-total=10000 --key-space-size=1000
Watch events received summary:
 Requests/sec: 45727.6095

# key-space-size=10
benchmark --conns=5 --clients=150 \
	watch --watchers=10000 --total=50000 --watched-key-total=10000 \
	--put-rate=1000 --put-total=10000 --key-space-size=10
Watch events received summary:
 Requests/sec: 95437.5692    =======> (not 100x, only 2x)
```

This could be affected by `put-rate` because higher-throughput means more data for watchers to process, which takes more time to receive events. So I ran the same tests with different `put-rate` to see the relationships:

```
# put-rate=10
# key-space-size=1000
Watch events received summary:
 Requests/sec: 99.7738

# put-rate=10
# key-space-size=10 (not 100x, only 50x)
Watch events received summary:
 Requests/sec: 4961.5690
```

```
# put-rate=1
# key-space-size=1000
Watch events received summary:
 Requests/sec: 1.1628

# put-rate=1
# key-space-size=10 (about 100x faster)
Watch events received summary:
 Requests/sec: 100.6869
```

This explains that watch performance can be affected by:
- how fast clients writes the keys that are being watched (`put-rate`)
   - the faster the events happen on the keys, the slower the clients receive events from the watch channel.
  - when clients write 1,000 keys with `1 put/sec` rate, the watch performance was not affected. For 10 different keys, it can process 100 requests per second.
  - when clients write 1,000 keys with `10 put/sec` rate, the watch performance was affected. But this doesn't mean watch is slow. For 1,000 different keys, it processes 100 requests per second. For 10 different keys, it processes 5,000 events per second. Ideally it should process 10,000 events per second compared to 1,000 different-key cases, but with higher throughput, etcd takes more time to process events.

In conclusion, watch clients should expect few seconds of extra event processing time when they write with rates higher than `1 PUT/sec`.

**Comment by xiang90** [comment]
OK. This makes sense. We will need to have more detailed analysis in the future. But this is enough for now.

**Comment by gyuho** [comment]
@xiang90 Did another run trying to get the maximum numbers:

```
benchmark --conns=5 --clients=150 \
	watch --watchers=200000 --total=50000 --watched-key-total=10000 \
	--put-rate=3000 --put-total=100000 --key-space-size=1 \
;
Watch events received summary:
 Requests/sec:  138350.4594

benchmark --conns=5 --clients=150 \
	watch --watchers=200000 --total=50000 --watched-key-total=10000 \
	--put-rate=3000 --put-total=100000 --key-space-size=1000 \
;
Watch events received summary:
 Requests/sec:  113396.9432
 ```

If we push more than this, etcd watch requests hang. So maximum watch request throughput would be 100,000 requests per second. This is only a rough number. If someone needs this much throughput, they need to run their own tests to their use case.

Thanks.

**Comment by xiang90** [comment]
OK. Closing for now. We will probably write some doc around watch performance in the future.

**Comment by gyuho** [comment]
Yeah I write a short doc in Google drive. Thanks.

**Comment by ivdimitro** [comment]
Hi, Have you investigated further the performance impact of watch in ETCD 2.x? Have you written this doc :)
